### ISSUE TITLE
services.ConfigureOptions<T> does not register validations

### ISSUE THREAD

**Issue by loic-sharma** [dotnet/runtime#38491]
### Description

The [`OptionsServiceCollectionExtensions.ConfigureOptions`](https://github.com/dotnet/runtime/blob/0ab928b27bb838a27d7712e0f87312a85615cbe5/src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs#L170) method is used to automatically register `I[Post]ConfigureOptions`. It would be convenient if it also registered `IValidateOptions`.

For example, say I have this class:

```csharp
public class ConfigureAppOptions : IConfigureOptions<AppOptions>, IValidateOptions<AppOptions>
{
  public void Configure(AppOptions options)
  {
    options.Foo = "Bar";
  }

  public ValidateOptionsResult Validate(string name, AppOptions options)
  {
    if (options.Foo != "Bar")
    {
      return ValidateOptionsResult.Fail("I find your lack of foo bar disturbing");
    }

    return ValidateOptions.Success;
  }
}
```

Now I register it in my dependency injection container:

```csharp
services.ConfigureOptions<ConfigureAppOptions>();
```

When I create `AppOptions`, the `Configure` method will run as expected... however, my `Validate` method won't! 😢

### Workaround

To fix this, I need to register my validation manually:

```csharp
services.ConfigureOptions<ConfigureAppOptions>();
services.AddTransient<IValidateOptions<AppOptions>, ConfigureAppOptions>());
```

### Real world example

See:
* https://github.com/loic-sharma/BaGet/blob/872fd2e92aaf055eb144975a8572b5337cc45f27/src/BaGet/Startup.cs#L30-L41
* https://github.com/loic-sharma/BaGet/blob/872fd2e92aaf055eb144975a8572b5337cc45f27/src/BaGet/ConfigureBaGetOptions.cs#L17-L22

### Fix

The [`FindIConfigureOptions` method](https://github.com/dotnet/runtime/blob/0ab928b27bb838a27d7712e0f87312a85615cbe5/src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs#L148-L162) will likely need to be updated:

```diff
private static IEnumerable<Type> FindIConfigureOptions(Type type)
{
    IEnumerable<Type> serviceTypes = type.GetTypeInfo().ImplementedInterfaces
        .Where(t => t.GetTypeInfo().IsGenericType &&
        (t.GetGenericTypeDefinition() == typeof(IConfigureOptions<>)
        || t.GetGenericTypeDefinition() == typeof(IPostConfigureOptions<>)
+       || t.GetGenericTypeDefinition() == typeof(IValidateOptions<>)));
    if (!serviceTypes.Any())
    {
        throw new InvalidOperationException(
            IsAction(type)
            ? SR.Error_NoIConfigureOptionsAndAction
            : SR.Error_NoIConfigureOptions);
    }
    return serviceTypes;
}
```

Likewise, the error strings will likely need to be updated ([link](https://github.com/dotnet/runtime/blob/473d6f9de26db5f8442bb71c9744c01de58f7d09/src/libraries/Microsoft.Extensions.Options/src/Resources/Strings.resx#L132-L137)).

I'd be happy to contribute this fix if y'all like it.

**Comment by HaoK** [comment]
Yeah for sure, that seems like a great idea to me, my team doesn't own this feature area anymore, but your proposed fix is definitely what the behavior should be

**Comment by HaoK** [comment]
The only small issue I have is naming of the method is a ConfigureOptions, but since this one is doing reflection magic and currently takes care of IPostConfigure anyways, it still seems reasonable to have it also do IValidateOptions

**Comment by loic-sharma** [comment]
> The only small issue I have is naming of the method is a ConfigureOptions, but since this one is doing reflection magic and currently takes care of IPostConfigure anyways, it still seems reasonable to have it also do IValidateOptions

I agree. What do you think of the following renames?

* `FindIConfigureOptions` to `FindConfigurationServices`
* `SR.Error_NoIConfigureOptions` to `SR.Error_NoConfigurationServices`
* `SR.Error_NoIConfigureOptionsAndAction` to `SR.Error_NoConfigurationServicesAndAction`

**Comment by loic-sharma** [comment]
> The only small issue I have is naming of the method is a ConfigureOptions, but since this one is doing reflection magic and currently takes care of IPostConfigure anyways, it still seems reasonable to have it also do IValidateOptions

Agreed. Here are some options:

1. Update `ConfigureOptions` to register `IValidateOptions`. This may break customer's apps that already call `ConfigureOptions` on a type that also implements `IValidateOptions<>`. It's not particularly clear from the name whether this will register validations, but, I think updating the comment could address this.
1. Don't register validations, but, update the comment on `ConfigureOptions` to specify that validations will not be registered.
1. Add an overload to `ConfigureOptions` to specify whether we should register validations. Something like `ConfigureServices<T>(addValidations: true);`. This is a non-breaking change. This may confuse users that assume `ConfigureOptions` will register validations by default.
1. Add a new method like `ConfigureAndValidateOptions` or `Configure. This is a non-breaking change. This may confuse users that assume `ConfigureOptions` will register validations by default.
1. Rename `ConfigureOptions`. I'm guessing this is not an option.

Personally I prefer option 1, followed by option 2. Do you agree? Do you have any other suggestions? Should we review this proposal with stakeholders, and if so, who are they? Thanks for your help!

**Comment by HaoK** [comment]
Right, I think option 1 is probably okay, to just update the comments and explicitly state validations will be registered by `ConfigureOptions`, this is 5.0 so not the worst time to introduce a small breaking change.  I find it hard to think of many situations where someone would implement `IValidateOptions` on this class and not want them registered, especially since specific named instances can targeted

**Comment by HaoK** [comment]
I think its best to just wait for the normal triage process for the next steps, but never hurts to see if @davidfowl agrees with our thinking

**Comment by HaoK** [comment]
BTW just as an FYI, you can also do your registration via `OptionsBuilder` without the class

```
  services.AddOptions<AppOptions>
     .Configure(options => options.Foo = "Bar")
     .Validate(name, options => {
        if (options.Foo != "Bar")
        {
          return false;
        }
        return true
     }, "I find your lack of foo bar disturbing");
```

**Comment by loic-sharma** [comment]
I sent out a pull request with the current proposal: https://github.com/dotnet/runtime/pull/38513. Please let me know if you have any feedback!

**Comment by maryamariyan** [comment]
@HaoK thanks for the feedback thread.

**Comment by loic-sharma** [comment]
@maryamariyan is the up-for-grabs label correct? The PR for this issue has been merged already: https://github.com/dotnet/runtime/pull/38513

**Comment by eriawan** [comment]
@maryamariyan 
the PR that fixes this issue has been merged: #38513

Could we close this issue

**Comment by maryamariyan** [comment]
> the PR that fixes this issue has been merged: #38513. Could we close this issue?

Sure. Closing as it's already fixed.
